components/tidb_query/src/codec/row/v2/encoder.rs

```diff
@@ -0,0 +1,295 @@
+        } else {
+            self.write_u16_le(offset as u16)
+        }
+    }
+}
+impl RowEncoder for Vec<u8> {}
+
+trait DatumEncoder: NumberEncoder + DecimalEncoder + JsonEncoder {
+    #[inline]
+    fn write_datum(&mut self, datum: Datum) -> Result<()> {
+        match datum {
+            Datum::I64(i) => self.encode_i64(i).map_err(Error::from),
+            Datum::U64(u) => self.encode_u64(u).map_err(Error::from),
+            Datum::Bytes(b) => self.write_bytes(&b).map_err(Error::from),
+            Datum::Dec(d) => {
+                let (prec, frac) = d.prec_and_frac();
+                self.write_decimal(&d, prec, frac)?;
+                Ok(())
+            }
+            Datum::Json(j) => self.write_json(&j),
+            Datum::F64(_) | Datum::Time(_) | Datum::Dur(_) | Datum::Null => unreachable!(),
+            Datum::Min | Datum::Max => unimplemented!("not supported yet"),
+        }
+    }
+
+    #[allow(clippy::match_overlapping_arm)]
+    #[inline]
+    fn encode_i64(&mut self, v: i64) -> codec::Result<()> {
+        match v {
+            i8::MIN_I64..=i8::MAX_I64 => self.write_u8(v as i8 as u8),
+            i16::MIN_I64..=i16::MAX_I64 => self.write_i16_le(v as i16),
+            i32::MIN_I64..=i32::MAX_I64 => self.write_i32_le(v as i32),
+            _ => self.write_i64_le(v),
+        }
+    }
+
+    #[allow(clippy::match_overlapping_arm)]
+    #[inline]
+    fn encode_u64(&mut self, v: u64) -> codec::Result<()> {
+        match v {
+            u8::MIN_U64..=u8::MAX_U64 => self.write_u8(v as u8),
+            u16::MIN_U64..=u16::MAX_U64 => self.write_u16_le(v as u16),
+            u32::MIN_U64..=u32::MAX_U64 => self.write_u32_le(v as u32),
+            _ => self.write_u64_le(v),
+        }
+    }
+}
+impl DatumEncoder for Vec<u8> {}
+
+struct Row {
+    is_big: bool,
+    non_null_cols: Vec<(Datum, i64)>,
+    null_cols: Vec<(Datum, i64)>,
+    value_approx_size: usize,
+}
+
+impl Row {
+    fn new(datums: Vec<Datum>, ids: &[i64]) -> Row {
+        let mut is_big = false;
+
+        let mut non_null_cols = Vec::with_capacity(ids.len());
+        let mut null_cols = Vec::with_capacity(ids.len());
+
+        let mut value_approx_size = 0;
+        for (datum, id) in datums.into_iter().zip(ids) {
+            let datum = flatten(datum);
+            if *id > 255 {
+                is_big = true;
+            }
+            value_approx_size += datum_size(&datum);
+
+            if datum == Datum::Null {
+                null_cols.push((datum, *id));
+            } else {
+                non_null_cols.push((datum, *id));
+            }
+        }
+        non_null_cols.sort_by_key(|(_, id)| *id);
+        null_cols.sort_by_key(|(_, id)| *id);
+
+        Row {
+            is_big,
+            non_null_cols,
+            null_cols,
+            value_approx_size,
+        }
+    }
+
+    // version(1) + flag(1) + columns_length(4) + ids_length + values_offset_length + values_length
+    #[inline]
```

Comment:
As a test only module function I prefer to remove these inlines and there is no need to estimate an approximate size. In the near future I believe we don't need the row encoder in the non-test code so it's ok to drop it :)

Reply:
done